File: src/benchmark.py
"""
benchmark.py — Run both solvers against all answers and compare.

This is the "proof" component — the result you cite on your resume.

Runs:
    - SolverV1 (brute-force / first candidate)
    - SolverV2 (entropy-optimal)

Outputs:
    - Per-word guess counts
    - Average, max, distribution
    - % solved within 6 guesses
    - Saves results/benchmark_results.csv

Usage:
    python benchmark.py [--quick]   # --quick runs on first 100 words only
"""

import os
import sys
import csv
import time
import argparse
from collections import Counter

# Add src to path
sys.path.insert(0, os.path.dirname(__file__))
from solver_v1 import SolverV1
from solver_v2 import SolverV2


def load_words(filepath: str) -> list[str]:
    """Load a word list from a plain text file (one word per line)."""
    with open(filepath) as f:
        return [w.strip().lower() for w in f if w.strip()]


def run_benchmark(solver, answers: list[str], name: str) -> dict:
    """
    Run `solver` against every word in `answers` and collect statistics.

    Args:
        solver:  Solver instance with a .solve(word) method.
        answers: List of all answer words to simulate.
        name:    Label for progress display.

    Returns:
        Dictionary with 'results' (list of guess counts) and summary stats.
    """
    results = []
    n = len(answers)
    start = time.time()

    print(f"\n{'='*50}")
    print(f" Running {name} on {n} words...")
    print(f"{'='*50}")

    for i, word in enumerate(answers):
        guesses = solver.solve(word, verbose=False)
        results.append(guesses)

        # Progress bar every 100 words
        if (i + 1) % 100 == 0 or (i + 1) == n:
            elapsed = time.time() - start
            rate = (i + 1) / elapsed
            eta = (n - i - 1) / rate if rate > 0 else 0
            bar_len = 30
            filled = int(bar_len * (i + 1) / n)
            bar = '█' * filled + '░' * (bar_len - filled)
            print(f"  [{bar}] {i+1}/{n} ({rate:.1f} w/s, ETA: {eta:.0f}s)", end='\r')

    elapsed = time.time() - start
    print(f"\n  Completed in {elapsed:.1f}s")

    # Compute summary stats
    avg = sum(results) / len(results)
    max_g = max(results)
    min_g = min(results)
    dist = Counter(results)
    within_6 = sum(1 for r in results if r <= 6) / len(results) * 100
    failed = sum(1 for r in results if r > 6)

    return {
        'name': name,
        'results': results,
        'avg': avg,
        'max': max_g,
        'min': min_g,
        'dist': dist,
        'within_6_pct': within_6,
        'failed': failed,
        'total': len(results),
        'time_s': elapsed,
    }


def print_summary(stats: dict):
    """Print a formatted summary for one solver's benchmark results."""
    print(f"\n  📊 Results for: {stats['name']}")
    print(f"  {'─'*40}")
    print(f"  Average guesses : {stats['avg']:.4f}")
    print(f"  Min guesses     : {stats['min']}")
    print(f"  Max guesses     : {stats['max']}")
    print(f"  Within 6 guesses: {stats['within_6_pct']:.2f}%  ({stats['total'] - stats['failed']}/{stats['total']})")
    print(f"  Failed (>6)     : {stats['failed']}")
    print(f"\n  Guess distribution:")
    for k in sorted(stats['dist'].keys()):
        count = stats['dist'][k]
        bar = '▓' * (count // 10)
        label = "FAIL" if k == 7 else str(k)
        print(f"    {label} guesses: {count:4d}  {bar}")


def save_csv(all_stats: list[dict], answers: list[str], out_path: str):
    """Save per-word results to a CSV file for further analysis."""
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    
    headers = ['word'] + [s['name'] for s in all_stats]
    with open(out_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        for i, word in enumerate(answers):
            row = [word] + [s['results'][i] for s in all_stats]
            writer.writerow(row)
    print(f"\n  💾 Results saved to: {out_path}")


def print_comparison(v1_stats: dict, v2_stats: dict):
    """Print a side-by-side comparison table."""
    print("\n" + "=" * 55)
    print(f"  {'COMPARISON TABLE':^51}")
    print("=" * 55)
    print(f"  {'Metric':<25} {'V1 (Brute Force)':>12} {'V2 (Entropy)':>12}")
    print(f"  {'-'*25} {'-'*12} {'-'*12}")
    print(f"  {'Average guesses':<25} {v1_stats['avg']:>12.4f} {v2_stats['avg']:>12.4f}")
    print(f"  {'Max guesses':<25} {v1_stats['max']:>12} {v2_stats['max']:>12}")
    print(f"  {'Solved within 6 (%)':<25} {v1_stats['within_6_pct']:>11.2f}% {v2_stats['within_6_pct']:>11.2f}%")
    print(f"  {'Failed (>6 guesses)':<25} {v1_stats['failed']:>12} {v2_stats['failed']:>12}")
    print(f"  {'Runtime':<25} {v1_stats['time_s']:>11.1f}s {v2_stats['time_s']:>11.1f}s")
    print("=" * 55)
    improvement = v1_stats['avg'] - v2_stats['avg']
    print(f"\n  🔥 Entropy solver saves {improvement:.2f} avg guesses vs brute force!")


def main():
    parser = argparse.ArgumentParser(description="Wordle Solver Benchmark")
    parser.add_argument('--quick', action='store_true',
                        help='Quick mode: only run first 100 words (for testing)')
    parser.add_argument('--v2-only', action='store_true',
                        help='Skip V1 benchmark (much faster)')
    args = parser.parse_args()

    # File paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    data_dir = os.path.join(script_dir, '..', 'data')
    results_dir = os.path.join(script_dir, '..', 'results')

    # Load word lists
    answers_path = os.path.join(data_dir, 'answers.txt')
    print(f"Loading words from: {answers_path}")
    answers = load_words(answers_path)

    if args.quick:
        answers = answers[:100]
        print(f"⚡ Quick mode: using first {len(answers)} words")

    print(f"Loaded {len(answers)} answer words")

    all_stats = []

    # --- Run V1 ---
    if not args.v2_only:
        v1 = SolverV1(answers)
        v1_stats = run_benchmark(v1, answers, "V1 — Brute Force")
        print_summary(v1_stats)
        all_stats.append(v1_stats)

    # --- Run V2 ---
    v2 = SolverV2(answers)
    v2_stats = run_benchmark(v2, answers, "V2 — Entropy Optimal")
    print_summary(v2_stats)
    all_stats.append(v2_stats)

    # --- Comparison ---
    if not args.v2_only:
        print_comparison(v1_stats, v2_stats)

    # --- Save results ---
    csv_path = os.path.join(results_dir, 'benchmark_results.csv')
    save_csv(all_stats, answers, csv_path)

    print("\n✅ Benchmark complete!")


if __name__ == "__main__":
    main()

File: src/benchmark.cpp
/**
 * benchmark.cpp — Run V1 and V2 solvers against all answer words and compare.
 *
 * Usage:
 *   ./benchmark                  # Full run (both V1 + V2, all 2315 words)
 *   ./benchmark --v2-only        # Skip V1 (faster)
 *   ./benchmark --quick          # First 100 words only (for quick testing)
 *
 * Output:
 *   - Per-solver stats: avg, max, distribution, % within 6
 *   - Side-by-side comparison table
 *   - results/benchmark_results.csv
 *
 * Expected results (C++ is ~50x faster than Python):
 *   V2: avg ~3.55 guesses, 99.48% within 6, runtime ~0.5s
 *   V1: avg ~4.80 guesses, ~92% within 6
 */

#include "feedback.h"
#include "solver.h"
#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <map>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <cstring>

// -----------------------------------------------------------------------
// Progress bar helper
// -----------------------------------------------------------------------
void print_progress(int done, int total, double rate, double eta) {
    constexpr int BAR_LEN = 30;
    int filled = (total > 0) ? (BAR_LEN * done / total) : 0;
    std::cout << "\r  [";
    for (int i = 0; i < BAR_LEN; i++)
        std::cout << (i < filled ? '\xE2' : '\xE2');  // Use simple chars for portability
    // Simpler ASCII progress bar
    std::cout << "\r  [";
    for (int i = 0; i < BAR_LEN; i++)
        std::cout << (i < filled ? '#' : '.');
    std::cout << "] " << done << "/" << total
              << " (" << std::fixed << std::setprecision(1) << rate << " w/s"
              << ", ETA: " << static_cast<int>(eta) << "s)  " << std::flush;
}

// -----------------------------------------------------------------------
// Benchmark runner
// -----------------------------------------------------------------------
struct BenchStats {
    std::string name;
    std::vector<int> results;
    double avg = 0;
    int    max_g = 0, min_g = 0;
    double within_6_pct = 0;
    int    failed = 0;
    double time_s = 0;
};

BenchStats run_benchmark(
    const std::string& name,
    const std::vector<std::string>& answers,
    int solver_version)  // 1 = V1, 2 = V2
{
    BenchStats stats;
    stats.name = name;
    stats.results.reserve(answers.size());

    int n = static_cast<int>(answers.size());
    auto start = std::chrono::steady_clock::now();

    std::cout << "\n" << std::string(52, '=') << "\n";
    std::cout << " Running " << name << " on " << n << " words...\n";
    std::cout << std::string(52, '=') << "\n";

    for (int i = 0; i < n; i++) {
        int guesses = (solver_version == 1)
                      ? solve_v1(answers[i], answers, false)
                      : solve_v2(answers[i], answers, false);
        stats.results.push_back(guesses);

        // Progress bar every 100 words
        if ((i + 1) % 100 == 0 || (i + 1) == n) {
            auto now     = std::chrono::steady_clock::now();
            double elapsed = std::chrono::duration<double>(now - start).count();
            double rate  = (i + 1) / elapsed;
            double eta   = (n - i - 1) / rate;
            print_progress(i + 1, n, rate, eta);
        }
    }
    std::cout << "\n";

    auto end = std::chrono::steady_clock::now();
    stats.time_s = std::chrono::duration<double>(end - start).count();

    // Compute statistics
    double sum = 0;
    stats.min_g = stats.results[0];
    stats.max_g = stats.results[0];
    for (int g : stats.results) {
        sum += g;
        if (g > stats.max_g) stats.max_g = g;
        if (g < stats.min_g) stats.min_g = g;
        if (g > 6) stats.failed++;
    }
    stats.avg = sum / n;
    stats.within_6_pct = 100.0 * (n - stats.failed) / n;

    return stats;
}

// -----------------------------------------------------------------------
// Print helpers
// -----------------------------------------------------------------------
void print_summary(const BenchStats& s) {
    std::map<int, int> dist;
    for (int g : s.results) dist[g]++;

    std::cout << "\n  Results for: " << s.name << "\n";
    std::cout << "  " << std::string(42, '-') << "\n";
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "  Average guesses : " << s.avg << "\n";
    std::cout << "  Min guesses     : " << s.min_g << "\n";
    std::cout << "  Max guesses     : " << s.max_g << "\n";
    std::cout << std::setprecision(2);
    std::cout << "  Within 6 guesses: " << s.within_6_pct << "%"
              << "  (" << (s.results.size() - s.failed) << "/" << s.results.size() << ")\n";
    std::cout << "  Failed (>6)     : " << s.failed << "\n";
    std::cout << "\n  Guess distribution:\n";

    for (auto& [k, count] : dist) {
        std::string bar(count / 10, '#');
        std::cout << "    " << (k == 7 ? "FAIL" : std::to_string(k))
                  << " guesses: " << std::setw(5) << count
                  << "  " << bar << "\n";
    }
}

void print_comparison(const BenchStats& v1, const BenchStats& v2) {
    std::cout << "\n" << std::string(57, '=') << "\n";
    std::cout << "              COMPARISON TABLE\n";
    std::cout << std::string(57, '=') << "\n";
    std::cout << std::left << std::setw(26) << "  Metric"
              << std::right << std::setw(14) << "V1 (Brute)"
              << std::setw(14) << "V2 (Entropy)" << "\n";
    std::cout << "  " << std::string(53, '-') << "\n";

    auto row = [&](const std::string& label, auto a, auto b) {
        std::cout << std::left << std::setw(26) << ("  " + label)
                  << std::right << std::setw(14) << a
                  << std::setw(14) << b << "\n";
    };

    std::cout << std::fixed << std::setprecision(4);
    row("Average guesses",    v1.avg,          v2.avg);
    std::cout << std::setprecision(0);
    row("Max guesses",        v1.max_g,        v2.max_g);
    std::cout << std::setprecision(2);
    row("Solved within 6 (%)", v1.within_6_pct, v2.within_6_pct);
    row("Failed (>6)",        v1.failed,       v2.failed);
    std::cout << std::setprecision(1);
    row("Runtime (s)",        v1.time_s,       v2.time_s);
    std::cout << std::string(57, '=') << "\n";
    std::cout << std::setprecision(4);
    std::cout << "\n  Entropy solver saves "
              << (v1.avg - v2.avg) << " avg guesses vs brute force!\n";
}

void save_csv(const std::vector<BenchStats>& all_stats,
              const std::vector<std::string>& answers,
              const std::string& out_path)
{
    // Ensure directory exists
    std::filesystem::path p(out_path);
    std::filesystem::create_directories(p.parent_path());

    std::ofstream f(out_path);
    if (!f.is_open()) {
        std::cerr << "  Warning: could not write CSV to " << out_path << "\n";
        return;
    }

    // Header
    f << "word";
    for (const auto& s : all_stats) f << "," << s.name;
    f << "\n";

    // Rows
    for (size_t i = 0; i < answers.size(); i++) {
        f << answers[i];
        for (const auto& s : all_stats) f << "," << s.results[i];
        f << "\n";
    }
    std::cout << "\n  CSV saved to: " << out_path << "\n";
}

// -----------------------------------------------------------------------
// Main
// -----------------------------------------------------------------------
int main(int argc, char* argv[]) {
    bool quick   = false;
    bool v2_only = false;

    for (int i = 1; i < argc; i++) {
        if (std::strcmp(argv[i], "--quick")   == 0) quick   = true;
        if (std::strcmp(argv[i], "--v2-only") == 0) v2_only = true;
    }

    // Locate data directory relative to binary location
    std::string data_dir = "data/";  // Run from project root

    std::cout << "Loading word lists...\n";
    std::vector<std::string> answers;
    try {
        answers = load_words(data_dir + "answers.txt");
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        std::cerr << "Run from the project root: ./benchmark\n";
        return 1;
    }

    if (quick) {
        answers.resize(std::min(100, (int)answers.size()));
        std::cout << "Quick mode: using first " << answers.size() << " words\n";
    }

    std::cout << "Loaded " << answers.size() << " answer words\n";

    std::vector<BenchStats> all_stats;

    // --- V1 ---
    if (!v2_only) {
        auto s = run_benchmark("V1 -- Brute Force", answers, 1);
        print_summary(s);
        all_stats.push_back(s);
    }

    // --- V2 ---
    {
        auto s = run_benchmark("V2 -- Entropy Optimal", answers, 2);
        print_summary(s);
        all_stats.push_back(s);
    }

    // --- Comparison ---
    if (!v2_only && all_stats.size() == 2) {
        print_comparison(all_stats[0], all_stats[1]);
    }

    // --- Save CSV ---
    save_csv(all_stats, answers, "results/benchmark_results.csv");

    std::cout << "\nBenchmark complete!\n";
    return 0;
}
